components/scream/src/physics/p3/p3_functions_calc_rime_density_impl.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef P3_FUNCTIONS_CALC_RIME_DENSITY_IMPL_HPP
+#define P3_FUNCTIONS_CALC_RIME_DENSITY_IMPL_HPP
+
+#include "p3_functions.hpp" // for ETI only but harmless for GPU
+
+namespace scream {
+namespace p3 {
+
+/*
+ * Implementation of p3 rime density calculation function.
+ * Clients should NOT #include this file, but include p3_functions.hpp instead.
+ */
+
+template <typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::calc_rime_density(const Spack& t, const Spack& rhofaci,
+                    const Spack& f1pr02, const Spack& acn,
+                    const Spack& lamc, const Spack& mu_c,
+                    const Spack& qc_incld, const Spack& qccol,
+                    Spack& vtrmi1, Spack& rhorime_c)
+{
+  constexpr Scalar qsmall = C::QSMALL;
+  constexpr Scalar RainFrze = C::RainFrze;
+  constexpr Scalar ZeroDegC = C::ZeroDegC;
+  constexpr Scalar bcn = C::bcn;
+
+  const auto qccol_not_small_and_t_freezing = (qccol >= qsmall) &&
+                                              (t <= RainFrze);
+  const auto qccol_and_qc_not_small_and_t_freezing = (qccol >= qsmall) &&
+                                                     (qc_incld >= qsmall) &&
+                                                     (t <= RainFrze);
+
+  // NOTE: applicable for cloud only; modify when rain is added back
+  if (qccol_not_small_and_t_freezing.any()) {
+    // Mass-weighted mean ice fallspeed
+    vtrmi1.set(qccol_not_small_and_t_freezing, f1pr02 * rhofaci);
+
+    if (qccol_and_qc_not_small_and_t_freezing.any()) {
+
+      // Droplet fall speed (using Stokes' formulation, with analytic soln).
+      Spack Vt_qc;
+      Vt_qc.set(qccol_and_qc_not_small_and_t_freezing,
+                acn * tgamma(sp(4.0)+bcn+mu_c) /
+                (pow(lamc, bcn) * tgamma(sp(4.0)+mu_c)));
+
+      // Use mass-weighted mean size
+      Spack D_c, V_impact, inv_Tc, Ri;
+      D_c.set(qccol_and_qc_not_small_and_t_freezing, (sp(4.0) + mu_c) / lamc);
+      V_impact.set(qccol_and_qc_not_small_and_t_freezing, abs(vtrmi1-Vt_qc));
+      inv_Tc.set(qccol_and_qc_not_small_and_t_freezing,
+                 1/min(-0.001, t-ZeroDegC));
+      Ri.set(qccol_and_qc_not_small_and_t_freezing,
+             max(1, min(-0.5e+6 * D_c * V_impact * inv_Tc, sp(12.0))));
```

Comment:
This line looks mistranslated. Parens look  like this: `-(0.5e+6_rtype*D_c)*V_impact*iTc`. If the unit test isn't setting off klaxons, need to make it more sensitive.

Reply:
I know what you mean, but the scientific notation is misleading:
```
-(0.5e+6*D_c)*V_impact*iTc = -(500000*D_c)*V_impact*iTc
                           = -500000 * D_c * V_impact * iTc
                           = -0.5e+6 * D_c * V_impact * iTc
```

Reply:
You know, that said, we might still need those parens to get to BFB. I'm still seeing differences that are bigger than multiplicative non-commutativity, though (1e-7ish), so something else must still be going on.

Alternatively, I could just remove the parens from the original code since I'm arguably already altering it in a non-BFB by forcing rhorime_c to be set in all circumstances (per our discussion below). That would make it easier to read.